File: server/src/db/queries/roomMember.ts
import RoomMember from '../../models/RoomMember';
import User from '../../models/User';
import { Transaction } from 'sequelize';

export async function addMemberToRoom(
  roomId: number, 
  userId: number, 
  isAdmin: boolean = false,
  canGrantAdmin: boolean = false,
  transaction?: Transaction
) {
  return RoomMember.create({
    roomId,
    userId,
    isAdmin,
    canGrantAdmin
  }, { transaction });
}

export async function removeMemberFromRoom(roomId: number, userId: number, transaction?: Transaction) {
  return RoomMember.destroy({
    where: {
      roomId,
      userId
    },
    transaction
  });
}

export async function getRoomMember(roomId: number, userId: number) {
  return RoomMember.findOne({
    where: {
      roomId,
      userId
    }
  });
}

export async function setMemberOnline(roomId: number, userId: number, online: boolean) {
  return RoomMember.update({ online }, {
    where: {
      roomId,
      userId
    }
  });
}

export async function getOnlineUsers(roomId: number) {
  // join RoomMember with User and select only online users
  // return RoomMember and username， map to UserListItem
  const members = await RoomMember.findAll({
    where: {
      roomId,
      online: true
    },
    include: [{
      model: User,
      attributes: ['username']
    }]
  });
  return members.map((member) => ({
    id: member.userId,
    username: member.User?.username,
    online: member.online,
    isAdmin: member.isAdmin,
  }));
}

File: server/internal/database/room_member.go
package database

import (
	"sync-player-server/internal/models"

	"gorm.io/gorm"
)

// AddMemberToRoom adds a member to a room
func AddMemberToRoom(roomID, userID uint, isAdmin, canGrantAdmin bool, tx ...*gorm.DB) (*models.RoomMember, error) {
	db := getDB(tx...)

	member := &models.RoomMember{
		RoomID:        roomID,
		UserID:        userID,
		IsAdmin:       isAdmin,
		CanGrantAdmin: canGrantAdmin,
	}

	if err := db.Create(member).Error; err != nil {
		return nil, err
	}

	return member, nil
}

// RemoveMemberFromRoom removes a member from a room
func RemoveMemberFromRoom(roomID, userID uint, tx ...*gorm.DB) error {
	db := getDB(tx...)

	return db.Where("room_id = ? AND user_id = ?", roomID, userID).Delete(&models.RoomMember{}).Error
}

// GetRoomMember retrieves a room member
func GetRoomMember(roomID, userID uint) (*models.RoomMember, error) {
	var member models.RoomMember
	if err := DB.Where("room_id = ? AND user_id = ?", roomID, userID).First(&member).Error; err != nil {
		return nil, err
	}
	return &member, nil
}

// SetMemberOnline updates the online status of a member
func SetMemberOnline(roomID, userID uint, online bool) error {
	return DB.Model(&models.RoomMember{}).
		Where("room_id = ? AND user_id = ?", roomID, userID).
		Update("online", online).Error
}

// OnlineUser represents an online user with their info
type OnlineUser struct {
	ID       uint   `json:"id"`
	Username string `json:"username"`
	Online   bool   `json:"online"`
	IsAdmin  bool   `json:"isAdmin"`
}

// GetOnlineUsers retrieves all online users in a room
func GetOnlineUsers(roomID uint) ([]OnlineUser, error) {
	var members []models.RoomMember
	err := DB.Where("room_id = ? AND online = ?", roomID, true).
		Preload("User").
		Find(&members).Error

	if err != nil {
		return nil, err
	}

	users := make([]OnlineUser, 0, len(members))
	for _, member := range members {
		if member.User != nil {
			users = append(users, OnlineUser{
				ID:       member.UserID,
				Username: member.User.Username,
				Online:   member.Online,
				IsAdmin:  member.IsAdmin,
			})
		}
	}

	return users, nil
}
